src/lib/CheapReduction.hs

```diff
@@ -147,26 +151,54 @@ cheapReduceWithDeclsRec decls cont = case decls of
         extendSubst (b@>SubstVal x) $
           cheapReduceWithDeclsRec rest cont
 
-class CheaplyReducibleE (e::E) where
-  cheapReduceE :: CheapReducer m => e i -> m i o (Atom o)
-
-instance CheaplyReducibleE e => CheaplyReducibleE (Abs (Nest Decl) e) where
+cheapReduceName :: CheapReducer m => Name c o -> m i o (AtomSubstVal c o)
+cheapReduceName v =
+  withNameColorRep (getNameColor v)
+    lookupEnv v >>= \case
+      AtomNameBinding (LetBound (DeclBinding _ _ e)) -> case e of
+        -- We avoid synthesizing the dictionaries during the traversal
+        -- and only do that when cheap reduction is performed on the expr directly.
+        Op (SynthesizeDict _ _) -> stuck
+        _ -> (SubstVal <$> dropSubst (cheapReduceE e)) <|> stuck
+      _ -> stuck
+  where stuck = return $ Rename v
+
+class CheaplyReducibleE (e::E) (e'::E) where
+  cheapReduceE :: CheapReducer m => e i -> m i o (e' o)
+
+instance CheaplyReducibleE Atom Atom where
+  cheapReduceE :: forall m i o. CheapReducer m => Atom i -> m i o (Atom o)
+  cheapReduceE a = liftImmut do
+    a' <- substM a
+    substMap <- fmap M.fromList $ dropSubst $
+      forM (freeVarsList AtomNameRep a') \v -> (v,) <$> cheapReduceName v
+    let subst :: Subst AtomSubstVal o o =
+          newSubst \v -> case getNameColor v of
+            AtomNameRep -> substMap M.! v
+            _ -> Rename v
+    DB bindings <- getDB
+    return $ substE (toScope bindings, subst) a'
```

Comment:
@dougalm please take a look at this awful trick and let me know if you can think of anything better. This is basically working around the limitation of the `newSubst` which doesn't let you use an effectful function. It's sensible that it has to be pure, but I'm not sure what's a good way of doing what I do here.

On the other hand... what if I just stashed all the bindings and other state and run cheapReduceName in a fresh monad underneath the subst... WDYT?

Reply:
I see. You want to be able to monadically traverse the free names, is that right? Something like:

```haskell
traverseFreeAtomNames :: (SubstE AtomSubstVal e, Monad m)
                      => (AtomName n -> m (Atom n)) -> e n -> m (e n)
```

Actually, an earlier version of `SubstE` worked exactly like this, and we used the monadic traversal to do hoisting (with the `Either Err` monad). But we ended up doing the hoisting in a different way so there were no users of the monadic traversal anymore and I removed it. Maybe we should add it back? We could add it as a method of `SubstE` alongside the pure version. The pure version is just the special case with the identity monad, but it's important enough that I'm tempted to keep it separate for performance reasons.